Mathlib/Data/Bitvec/Lemmas.lean

```diff
@@ -1,173 +1,152 @@
-  rw [List.foldl, List.foldl]
-  -- generalize the accumulator of foldl
-  generalize h : 0 = x
-  conv in addLsb x b =>
-    rw [← h]
-  clear h
-  simp only [List.foldl_append, List.foldl_cons, List.foldl_nil]
-  induction' xs with x xs xs_ih generalizing x
-  · simp
-    unfold addLsb
-    simp [Nat.mul_succ]
-  · simp
-    apply xs_ih
-#align bitvec.to_nat_append Bitvec.toNat_append
-
--- Porting Note: the mathlib3port version of the proof was :
---  simp [bits_to_nat_to_list]
---  unfold bits_to_nat add_lsb List.foldl cond
---  simp [cond_to_bool_mod_two]
-theorem bits_toNat_decide (n : ℕ) : Bitvec.toNat (decide (n % 2 = 1) ::ᵥ Vector.nil) = n % 2 := by
-  simp only [bitsToNat_toList, Vector.toList_singleton, Vector.head_cons]
-  unfold bitsToNat addLsb List.foldl
-  simp [Nat.cond_decide_mod_two, -Bool.cond_decide]
-#align bitvec.bits_to_nat_to_bool Bitvec.bits_toNat_decide
-
-theorem ofNat_succ {k n : ℕ} :
-    Bitvec.ofNat k.succ n = Bitvec.ofNat k (n / 2)++ₜdecide (n % 2 = 1) ::ᵥ Vector.nil :=
+
+theorem toNat_append {msbs : BitVec w} {lsbs : BitVec v} :
+    (msbs ++ lsbs).toNat = msbs.toNat <<< v ||| lsbs.toNat := by
+  rcases msbs with ⟨msbs, hm⟩
+  rcases lsbs with ⟨lsbs, hl⟩
+  simp only [HAppend.hAppend, append, toNat_ofFin]
+  rw [toNat_ofNat (Nat.add_comm w v ▸ append_lt hl hm)]
+
+#noalign bitvec.bits_to_nat_to_bool
+#noalign bitvec.to_nat_of_nat
+
+@[simp]
+lemma extractLsb_eq {w : ℕ} (hi lo : ℕ) (a : BitVec w) :
+    extractLsb hi lo a = extractLsb' lo (hi - lo + 1) a :=
   rfl
-#align bitvec.of_nat_succ Bitvec.ofNat_succ
 
-theorem toNat_ofNat {k n : ℕ} : Bitvec.toNat (Bitvec.ofNat k n) = n % 2 ^ k := by
-  induction' k with k ih generalizing n
-  · simp [Nat.mod_one]
-    rfl
-  · rw [ofNat_succ, toNat_append, ih, bits_toNat_decide, mod_pow_succ, Nat.mul_comm]
-#align bitvec.to_nat_of_nat Bitvec.toNat_ofNat
+theorem toNat_extractLsb' {i j} {x : BitVec w} :
+    (extractLsb' i j x).toNat = x.toNat / 2 ^ i % (2 ^ j) := by
+  simp only [extractLsb', ofNat_eq_mod_two_pow, shiftRight_eq_div_pow]
+
+theorem getLsb_eq_testBit {i} {x : BitVec w} : getLsb x i = x.toNat.testBit i := by
+  simp only [getLsb, Nat.shiftLeft_eq, one_mul, Nat.and_two_pow]
+  cases' testBit (BitVec.toNat x) i
+  <;> simp [pos_iff_ne_zero.mp (two_pow_pos i)]
 
 theorem ofFin_val {n : ℕ} (i : Fin <| 2 ^ n) : (ofFin i).toNat = i.val := by
-  rw [ofFin, toNat_ofNat, Nat.mod_eq_of_lt]
-  apply i.is_lt
-#align bitvec.of_fin_val Bitvec.ofFin_val
+  rfl
+#align bitvec.of_fin_val Std.BitVec.ofFin_val
 
 theorem addLsb_eq_twice_add_one {x b} : addLsb x b = 2 * x + cond b 1 0 := by
   simp [addLsb, two_mul]
-#align bitvec.add_lsb_eq_twice_add_one Bitvec.addLsb_eq_twice_add_one
-
-theorem toNat_eq_foldr_reverse {n : ℕ} (v : Bitvec n) :
-    v.toNat = v.toList.reverse.foldr (flip addLsb) 0 := by rw [List.foldr_reverse]; rfl
-#align bitvec.to_nat_eq_foldr_reverse Bitvec.toNat_eq_foldr_reverse
```

Comment:
@eric-wieser, do you think this lemma is needed? Its a little hard for me to prove, but maybe someone else could try? Or we leave it out if not strictly necessary.

Reply:
I guess the reverse in the theorem name does not make sense anymore. I wonder if bits maybe returned the bits in the opposite order than toList?

Reply:
Indeed, the old `toList` returned the bits in big-endian order (where the head of the list is the MSB), whereas `Nat.bits` returns bits in little-endian order (the head of the list is the LSB).

Reply:
Should we then drop the reverse from the theorem name?

Reply:
I think this theorem as it is currently stated is quite different from the original.
The current statement:
```lean
v.toNat = v.toNat.bits.foldr bit 0
```
is really a theorem about `Nat`s, not `BitVec`s.

The equivalent would be something akin to
```lean
v.toNat = v.toBEList.reverse.foldr (flip addLsb) 0
```
Although this should be equivalent to  
```lean
v.toNat = v.toBEList.foldl (flip addLsb) 0
```

I'm not sure why the original theorem used `reverse` and `foldr` over `foldl`

Reply:
An update on this: to prove `v.toNat = v.toBEList.foldl (flip addLsb) 0` cleanly, we really want to be able to do induction on a bitvector in terms of a `cons` or `concat` operation. Adding this in the current PR seems undesirable, since that would make the PR even larger.
So, instead I propose we change course, rather than replace the entire `Bitvec` API in one go, let's keep both `Bitvec` and `Std.BitVec` types while incrementally adding results to the latter. I've opened #8345 as a first step in that direction.

Reply:
I removed this lemma for now and replaced it with #noalign in this PR. We can then decide if we want to merge this PR with noalign or instead go for #8345.

Reply:
I think merging with noaligns is fine